### PR TITLE
Predict size of hash table for GROUP BY

### LINKED ISSUE #24317: Predict size of hash table for GROUP BY
**Describe the situation**

For simple queries sometimes around 10% of query execution time can be spent in hash table resizes. We have already optimized resizes but we can do better avoiding resizes whenever possible.

**Proposed solution**

Add a statistics (cache) for observed hash table sizes during GROUP BY. It can be keyed by hash of database and table names, GROUP BY and WHERE expressions.

When aggregation is finished, observed size is collected. If observed size is bigger than some threshold (e.g. 10 million items, controlled by query setting), it will not be used in statistics - it is needed to avoid memory overallocation. If observed size is greater than previously collected value, the previously collected value is replaced by the observed value (statistics is collecting the maximum observed size unless it is too large). If observed size is more than two times less than predicted value, the previously collected value is replaced by the observed value (the statistics is dropped on misprediction).

Two-level aggregation method can be selected from the start if predicted size is large enough.

The size of the cache should be controlled by server setting. Proposed size is 10 000 items. The cache can use LRU strategy. The usage of the cache (update of statistics and the usage of collected data) should be controlled by dedicated query-level setting (on/off).

### PR DESCRIPTION
**Pull request by nickitat**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Sizes of hash tables used during aggregation now collected and used in later queries to avoid hash tables resizes.

Detailed description / Documentation draft:
...

I see the following options to improve the current implementation:
* For now statistics collecting and usage is not implemented for merging partially aggregated results. It happens in a notable number of queries (distributed, in-order, cube & rollup), so it may worth doing.

Tl;dr [results](https://github.com/ClickHouse/ClickHouse/pull/33439#issuecomment-1075546628)

Issue #24317

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> Maybe smth like SYSTEM DROP HASH TABLE STATISTICS should be provided by analogy with SYSTEM DROP * CACHE.

It looks very low level, maybe we can avoid adding this command.
I don't want that the users will use it :)

> The only "observable behaviour" here is performance and in theory performance tests on sufficiently rich set of queries may be enough. But there are enough places where things may go wrong and rely only on perf tests is insufficient IMO.

Just check that nothing broken - by our existing tests.
Also an option is to add ProfileEvents and then write a test that will check their values in system.query_log.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -483,6 +483,10 @@ class IColumn;
     M(Bool, optimize_rewrite_sum_if_to_count_if, true, "Rewrite sumIf() and sum(if()) function countIf() function when logically equivalent", 0) \
     M(UInt64, insert_shard_id, 0, "If non zero, when insert into a distributed table, the data will be inserted into the shard `insert_shard_id` synchronously. Possible values range from 1 to `shards_number` of corresponding distributed table", 0) \
     \
+    M(Bool, collect_hash_table_stats_during_aggregation, true, "Enable collecting hash table statistics to optimize memory allocation", 0) \
+    M(UInt64, max_entries_for_hash_table_stats, 10'000, "How much entries hash table statistics collected during aggregation is allowed to have", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:487]: much -> many.

File: tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+           WHERE event_date >= yesterday() AND query_id = {query_id:String} AND message ILIKE '%$1%' \
+        ) AS res                                                                                     \
+        SELECT res $2"
+}
+
+check_number_of_string_occurrences_on_peer_satisfies_condition() {
+  $CLICKHOUSE_CLIENT -q "SYSTEM FLUSH LOGS"
+  $CLICKHOUSE_CLIENT                                                                             \
+    --param_query_id="$query_id"                                                                 \
+    -q "WITH (                                                                                   \
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh:62]: It should work without newline escaping (inside string literal).

**Comment by alexey-milovidov** [comment]
BTW, what performance improvement do you observe with manual testing?

File: src/Interpreters/Aggregator.cpp [hunk 3]
```diff
@@ -26,6 +27,133 @@
...
+
+    CachePtr getHashTableStatsCache(const Params & params, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    {
+        if (!hash_table_stats || hash_table_stats->maxSize() != params.max_entries_for_hash_table_stats)
+            hash_table_stats = std::make_shared<Cache>(params.max_entries_for_hash_table_stats);
+        return hash_table_stats;
+    }
+
+    std::mutex mutex;
+    CachePtr hash_table_stats;
```
Comment (azat) [inline src/Interpreters/Aggregator.cpp:78]: Worth exporting this LRUCache in the same manner as others, in AsynchronousMetrics.cpp:

https://github.com/ClickHouse/ClickHouse/blob/b1d2bdf5699c3726dc7f6fd3bca744eb267c187a/src/Interpreters/AsynchronousMetrics.cpp#L545-L551

**Comment by nickitat** [comment]
> BTW, what performance improvement do you observe with manual testing?

Here are some numbers:
```
display-name    0       SELECT number FROM numbers(10000000) GROUP BY number FORMAT Null SETTINGS collect_hash_table_stats_during_aggregation = 0
prewarm 0       hash_table_sizes_stats.query0.prewarm0  0 0.7254605293273926
query   0       hash_table_sizes_stats.query0.run0      0 0.724083662033081
query   0       hash_table_sizes_stats.query0.run1      0 0.7410163879394531
query   0       hash_table_sizes_stats.query0.run2      0 0.7370643615722656
query   0       hash_table_sizes_stats.query0.run3      0 0.7427923679351807
query   0       hash_table_sizes_stats.query0.run4      0 0.7495505809783936
query   0       hash_table_sizes_stats.query0.run5      0 0.7057178020477295
query   0       hash_table_sizes_stats.query0.run6      0 0.7060608863830566
client-time     0       5.106584181998187       5.10628604888916
display-name    1       SELECT number FROM numbers(10000000) GROUP BY number FORMAT Null SETTINGS collect_hash_table_stats_during_aggregation = 1
prewarm 1       hash_table_sizes_stats.query1.prewarm0  0 0.5177068710327148
query   1       hash_table_sizes_stats.query1.run0      0 0.5138120651245117
query   1       hash_table_sizes_stats.query1.run1      0 0.516057014465332
query   1       hash_table_sizes_stats.query1.run2      0 0.511322021484375
query   1       hash_table_sizes_stats.query1.run3      0 0.5145668983459473
query   1       hash_table_sizes_stats.query1.run4      0 0.5127284526824951
query   1       hash_table_sizes_stats.query1.run5      0 0.5113763809204102
query   1       hash_table_sizes_stats.query1.run6      0 0.5138087272644043
client-time     1       3.593963317998714       3.5936715602874756
display-name    2       SELECT number FROM numbers_mt(500000) GROUP BY number format Null SETTINGS collect_hash_table_stats_during_aggregation = 0
prewarm 2       hash_table_sizes_stats.query2.prewarm0  0 0.03299689292907715
query   2       hash_table_sizes_stats.query2.run0      0 0.030817031860351562
query   2       hash_table_sizes_stats.query2.run1      0 0.030933618545532227
query   2       hash_table_sizes_stats.query2.run2      0 0.03640604019165039
query   2       hash_table_sizes_stats.query2.run3      0 0.026384830474853516
query   2       hash_table_sizes_stats.query2.run4      0 0.03062891960144043
query   2       hash_table_sizes_stats.query2.run5      0 0.025572776794433594
query   2       hash_table_sizes_stats.query2.run6      0 0.02617502212524414
client-time     2       0.20719847900181776     0.20691823959350586
display-name    3       SELECT number FROM numbers_mt(500000) GROUP BY number format Null SETTINGS collect_hash_table_stats_during_aggregation = 1
prewarm 3       hash_table_sizes_stats.query3.prewarm0  0 0.015240907669067383
query   3       hash_table_sizes_stats.query3.run0      0 0.01188802719116211
query   3       hash_table_sizes_stats.query3.run1      0 0.012000560760498047
query   3       hash_table_sizes_stats.query3.run2      0 0.0118408203125
query   3       hash_table_sizes_stats.query3.run3      0 0.0110015869140625
query   3       hash_table_sizes_stats.query3.run4      0 0.012171268463134766
query   3       hash_table_sizes_stats.query3.run5      0 0.014752864837646484
query   3       hash_table_sizes_stats.query3.run6      0 0.011242151260375977
client-time     3       0.08522102699862444     0.08489727973937988
display-name    4       SELECT number FROM numbers_mt(1000000) GROUP BY number format Null SETTINGS collect_hash_table_stats_during_aggregation = 0
prewarm 4       hash_table_sizes_stats.query4.prewarm0  0 0.04354548454284668
query   4       hash_table_sizes_stats.query4.run0      0 0.03917956352233887
query   4       hash_table_sizes_stats.query4.run1      0 0.04040980339050293
query   4       hash_table_sizes_stats.query4.run2      0 0.03927874565124512
query   4       hash_table_sizes_stats.query4.run3      0 0.03807783126831055
query   4       hash_table_sizes_stats.query4.run4      0 0.04324507713317871
query   4       hash_table_sizes_stats.query4.run5      0 0.03917741775512695
query   4       hash_table_sizes_stats.query4.run6      0 0.04053497314453125
client-time     4       0.2802344739975524      0.2799034118652344
display-name    5       SELECT number FROM numbers_mt(1000000) GROUP BY number format Null SETTINGS collect_hash_table_stats_during_aggregation = 1
prewarm 5       hash_table_sizes_stats.query5.prewarm0  0 0.032332420349121094
query   5       hash_table_sizes_stats.query5.run0      0 0.033093929290771484
query   5       hash_table_sizes_stats.query5.run1      0 0.03340029716491699
query   5       hash_table_sizes_stats.query5.run2      0 0.03559160232543945
query   5       hash_table_sizes_stats.query5.run3      0 0.030649662017822266
query   5       hash_table_sizes_stats.query5.run4      0 0.03403615951538086
query   5       hash_table_sizes_stats.query5.run5      0 0.03227496147155762
query   5       hash_table_sizes_stats.query5.run6      0 0.030020475387573242
client-time     5       0.22936771600143402     0.22906708717346191
```
so atleast it performs good in single threaded case and in multi threaded when table size is such that before we decided to convert aggregation to two-level near to the end of the aggregation, but now can do that from the beginning. For bigger tables the profit of early convertion to two-level become less noticable.

Since the setting is enabled by default all existing perf tests are actually testing this logic, so I didn't spent much time on exploring different queries. And tests actually showed some improvements [one](https://s3.amazonaws.com/clickhouse-test-reports/33439/61e83292780272d547c1259c901bbbcc24a8e923/performance_comparison__actions__[2/4]/report.html), [two](https://s3.amazonaws.com/clickhouse-test-reports/33439/61e83292780272d547c1259c901bbbcc24a8e923/performance_comparison__actions__[3/4]/report.html), but also some [regression](https://s3.amazonaws.com/clickhouse-test-reports/33439/61e83292780272d547c1259c901bbbcc24a8e923/performance_comparison__actions__[4/4]/report.html) too, so there are something to look at. Mind to do that the first chance I'll have.

File: src/Interpreters/Aggregator.cpp (hunk 3 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:78]: thx, I replaced in the description the part about `SYSTEM DROP HASH TABLE STATISTICS` with what you suggested.

**Comment by UnamedRus** [comment]
> It looks very low level, maybe we can avoid adding this command.

In other DMBS like PostgreSQL, there is such commands which allow to alter table data statistics:
https://postgrespro.ru/docs/postgresql/10/sql-dropstatistics

**Comment by alexey-milovidov** [comment]
But this statistic is not about tables' data, it is something more ephemeral.

**Comment by nickitat** [comment]
Some queries from the "group_by_fixed_keys" test [have slowed down](https://s3.amazonaws.com/clickhouse-test-reports/33439/61e83292780272d547c1259c901bbbcc24a8e923/performance_comparison__actions__[4/4]/report.html). It happened because previously we used a hash table of 1024 elements (finally) and now this query will use HT of 512 elements (preallocated at the beginning). This is true for every second power of two, but seems like it makes things worse only for fairly small sizes (~4096 on my machine), but for bigger tables it actually gives us perf benefits (check the "Experiment results" section below). So I just left all this logic untouched.

<details>
  <summary>Experiment results</summary>

The following query from the "group_by_fixed_keys" was used:

```sql
    WITH number % SIZE AS k, toUInt64(k) AS k1, k1 + 1 AS k2
  SELECT k1, k2, count()
    FROM numbers(100000000)
GROUP BY k1, k2
  FORMAT Null
SETTINGS collect_hash_table_stats_during_aggregation = CS
```

I encourage you not to compare the values from different columns (but only from adjacent rows), since they are from different runs.

| \<SIZE\>_\<CS\>      | Current implementation | If we did like before |
| ----------- | ----------- | ----------- |
|**255_0**|**1.5808916091918945**|**1.5261054039001465**|
|**255_1**|**1.6315653324127197**|**1.552426815032959**|
|257_0|1.5843119621276855|1.5510237216949463|
|257_1|1.5761158466339111|1.555800199508667|
|511_0|1.7008326053619385|1.663393497467041|
|511_1|1.7077183723449707|1.6842105388641357|
|513_0|1.651308298110962|1.6336848735809326|
|513_1|1.6486756801605225|1.6197535991668701|
|**1023_0**|**1.7353456020355225**|**1.6944849491119385**|
|**1023_1**|**1.8785557746887207**|**1.6881804466247559**|
|1025_0|1.6594724655151367|1.6393749713897705|
|1025_1|1.6534018516540527|1.6477224826812744|
|2047_0|1.9330251216888428|1.8986682891845703|
|2047_1|1.944204568862915|1.9351270198822021|
|2049_0|1.7217907905578613|1.6737751960754395|
|2049_1|1.7120473384857178|1.685054063796997|
|**4095_0**|**1.9070887565612793**|**1.882392168045044**|
|**4095_1**|**2.0191121101379395**|**1.8864517211914062**|
|4097_0|1.8290636539459229|1.8191907405853271|
|4097_1|1.8279407024383545|1.7885808944702148|
|8191_0|1.9226725101470947|1.9338366985321045|
|8191_1|1.8923923969268799|1.8756773471832275|
|8193_0|1.902085542678833|1.8574156761169434|
|8193_1|1.8811531066894531|1.880134105682373|
|16383_0|2.0350911617279053|1.977320909500122|
|16383_1|2.0209827423095703|1.9731152057647705|
|16385_0|2.0228984355926514|1.933598518371582|
|16385_1|1.9846243858337402|1.9249930381774902|
|32767_0|2.1377968788146973|2.0998997688293457|
|32767_1|2.0927042961120605|2.088822841644287|
|**32769_0**|**3.0240345001220703**|**2.3491389751434326**|
|**32769_1**|**2.7204763889312744**|**2.344632625579834**|
|**65535_0**|**3.0606114864349365**|**2.3248038291931152**|
|**65535_1**|**2.7485265731811523**|**2.2993004322052**|
|65537_0|3.06917142868042|2.3265013694763184|
|65537_1|3.0379860401153564|2.350400447845459|
|131071_0|3.2479352951049805|2.5082695484161377|
|131071_1|3.236302137374878|2.424055576324463|
|**131073_0**|**6.623970985412598**|**5.846493721008301**|
|**131073_1**|**4.761634588241577**|**5.839216470718384**|
|**262143_0**|**6.754836082458496**|**6.059891700744629**|
|**262143_1**|**4.791410207748413**|**5.989444255828857**|
|262145_0|6.685712814331055|5.900468111038208|
|262145_1|6.6301140785217285|6.011543273925781|
|524287_0|6.935018301010132|6.260606050491333|
|524287_1|7.00163722038269|6.196463584899902|
|**524289_0**|**8.178144693374634**|**7.380904197692871**|
|**524289_1**|**7.439335346221924**|**7.432044744491577**|

Without preallocation we start with a hash table of `2^8` elements and then we make it 4 times bigger when size become equal to `2^(k+1)` for a next `k`. And this process is asymetric in a sense that, say, a query with `2^9+1` unique keys and a query with `2^10+1` keys will both deal with HT of `2^12` elements. With preallocation we will act more fair and a query with `2^9+1` keys will use HT of `2^11` elements. Effectively what happens is it lowers the amount of cache misses, e.g. for the `524289_0` vs `524289_1` I observed smth like `35.5%` vs `32%` in `perf stat` which pretty much the same ratio as `8.2s` vs `7.4s`.

For bigger sizes looks also OK (here the benefit of not doing reallocations also kicks in):

| \<SIZE\>_\<CS\>      | Current implementation |
| ----------- | ----------- |
|1048575_0|8.115857362747192|
|1048575_1|7.616235017776489|
|1048577_0|8.04812216758728|
|1048577_1|8.017758131027222|
|2097151_0|8.882615566253662|
|2097151_1|8.804941892623901|
|2097153_0|9.062370777130127|
|2097153_1|9.2320556640625|
|4194303_0|9.837358951568604|
|4194303_1|10.021861791610718|
|4194305_0|9.858987092971802|
|4194305_1|9.566895961761475|
|8388607_0|11.311129331588745|
|8388607_1|11.123503923416138|
|8388609_0|10.947338104248047|
|8388609_1|10.44078254699707|

</details>

Other slowdowns look irrelevant.

**Comment by nickitat** [comment]
Seems like smth happened with perf tests in CI, I see the same picture in other recent PRs.

File: src/Interpreters/Aggregator.cpp (hunk 3 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:92]: done

File: tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh (hunk 2 above)
Comment (nickitat) [reply tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh]: done

File: src/Core/Settings.h (hunk 1 above)
Comment (nickitat) [reply src/Core/Settings.h]: done

**Comment by alexey-milovidov** [comment]
There are errors in perf tests in master, @Avogar will fix.

**Comment by alexey-milovidov** [comment]
Maybe @vdimir can take.
If not, I will try to look.

PS. https://gh-api.clickhouse.tech/play?user=play#U0VMRUNUCiAgICBhY3Rvcl9sb2dpbiwKICAgIGNvdW50KCkgQVMgYywKICAgIGMgPj0gZGF0ZURpZmYoJ3dlZWsnLCAnMjAyMi0wMS0xMCc6OkRhdGUsIHRvZGF5KCkpICogNCBBUyBvawpGUk9NIGRlZmF1bHQuZ2l0aHViX2V2ZW50cwpXSEVSRSBjcmVhdGVkX2F0ID49ICcyMDIyLTAxLTAxJwogIEFORCByZXBvX25hbWUgPSAnQ2xpY2tIb3VzZS9DbGlja0hvdXNlJwogIEFORCBldmVudF90eXBlID0gJ1B1bGxSZXF1ZXN0RXZlbnQnCiAgQU5EIGFjdGlvbiA9ICdjbG9zZWQnCiAgQU5EIGFjdG9yX2xvZ2luID0gYXNzaWduZWUKR1JPVVAgQlkgYWN0b3JfbG9naW4KT1JERVIgQlkgYyBERVNDCg==

**Comment by nickitat** [comment]
CI spotted couple of small problems in tests. I will fix them soon.

File: src/Interpreters/Aggregator.cpp [hunk 4]
```diff
@@ -237,8 +469,7 @@ class CompiledAggregateFunctionsHolder final : public CompiledExpressionCacheEnt
 
 #endif
 
-Aggregator::Aggregator(const Params & params_)
-    : params(params_)
+Aggregator::Aggregator(const Params & params_) : params(params_), stats_updater(std::make_unique<StatisticsUpdater>(*this))
```
Comment (vdimir) [inline src/Interpreters/Aggregator.cpp:472]: Passing `this` from ctor looks dangerous because object is not fully constructed yet, but seems in this case it's fine, we will just store refernce to this somewhere

https://isocpp.org/wiki/faq/ctors#using-this-in-ctors

File: src/Interpreters/Aggregator.cpp [hunk 5]
```diff
@@ -27,12 +28,150 @@
...
+        return std::nullopt;
+    }
+
+    void update(size_t observed_size, const Params & params)
+    {
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
+        const auto hint = cache->get(params.key);
+        // We'll maintain the maximum among all the observed values until the next prediction turns out to be too wrong.
+        if (!hint || observed_size > *hint || observed_size < *hint / 2)
```
Comment (vdimir) [inline src/Interpreters/Aggregator.cpp:71]: It's really minor thing, but I suppose it's easier to read it like that, because we immediately see range when condition is not apllied `[hint/2, hint]`

```suggestion
        if (!hint || observed_size < *hint / 2 || *hint < observed_size)
```

File: src/Interpreters/Aggregator.cpp [hunk 6]
```diff
@@ -27,12 +28,150 @@
...
+        {
+            LOG_DEBUG(&Poco::Logger::get("Aggregator"), "Statistics updated for key={}: new size_hint={}", params.key, observed_size);
+            cache->set(params.key, std::make_shared<size_t>(observed_size));
+        }
+    }
+
+    CachePtr getCache()
```
Comment (vdimir) [inline src/Interpreters/Aggregator.cpp:78]: It's used only for `getStats`, so maybe let's expose `getStats` method here and return `HashTablesCacheStatistics`

File: src/Interpreters/Aggregator.h [hunk 7]
```diff
@@ -912,29 +913,60 @@ class Aggregator final
         bool compile_aggregate_expressions;
         size_t min_count_to_compile_aggregate_expression;
 
+        struct StatsCollectingParams
+        {
+            StatsCollectingParams();
```
Comment (vdimir) [inline src/Interpreters/Aggregator.h:918]: What's default contructor used for? I suppose we can provide more gurantees that we won't get inconsistent StatsCollectingParams, if we either:
* remove default ctor,
* use `key != 0` instead of `collect_hash_table_stats_during_aggregation` (it can't be `collect_hash_table_stats_during_aggregation = true` and `key = 0` at the same time, isn't it? In this case `collect_hash_table_stats_during_aggregation` may be reucntant)
* make fileds const to be sure that we construct only valid params
* Or even create `StatsCollectingParams` only when `collect_hash_table_stats_during_aggregation = true`, otherwise leave it nullptr

File: src/Interpreters/Aggregator.cpp (hunk 4 above)
Comment (vdimir) [reply src/Interpreters/Aggregator.cpp:472]: Also, maybe not to construct `stats_updater` at all if statistics in disabled, just leave `nullptr`?

File: src/Interpreters/Aggregator.cpp [hunk 8]
```diff
@@ -27,12 +28,150 @@
...
+    {
+        std::lock_guard lock(mutex);
+        return hash_table_stats;
+    }
+
+private:
+    CachePtr getHashTableStatsCache(const Params & params, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
```
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp]: yep, but i removed that while making other changes

File: src/Interpreters/Aggregator.cpp [hunk 9]
```diff
@@ -88,6 +222,104 @@ void AggregatedDataVariants::convertToTwoLevel()
...
+    bool collect_hash_table_stats_during_aggregation_,
+    size_t max_entries_for_hash_table_stats_,
+    size_t max_size_to_preallocate_for_aggregation_)
+    : key(0)
+    , collect_hash_table_stats_during_aggregation(collect_hash_table_stats_during_aggregation_)
+    , max_entries_for_hash_table_stats(max_entries_for_hash_table_stats_)
+    , max_size_to_preallocate_for_aggregation(max_size_to_preallocate_for_aggregation_)
+{
+    if (!select_query_)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "query ptr cannot be null");
```
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp]: it is being checked in `select_query->as<DB::ASTSelectQuery &>()`

File: src/Interpreters/Aggregator.cpp (hunk 4 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp]: > Also, maybe not to construct stats_updater at all if statistics in disabled, just leave nullptr?

done

File: src/Interpreters/Aggregator.cpp (hunk 5 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp]: done

File: src/Interpreters/Aggregator.cpp (hunk 6 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp]: done

File: src/Interpreters/Aggregator.h (hunk 7 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.h:921]: default ctor is used in those calls where we don't pass any statistics collection related params.

> use key != 0 instead of collect_hash_table_stats_during_aggregation

done

**Comment by nickitat** [comment]
FunctionalStatelessTestFlakyCheck failed because of timeout - 300 long tests haven't fit in timeout. but there were no fails, so I think we can ignore that.

**Comment by nickitat** [comment]
> Actually eleminating resizes on ordinary merging phase seems to give another ~10% for numbers_mt(10M) and ~20% for numbers_mt(50M) (i.e. for aggregation by high-cardinality columns)

I want to work on it a little

File: src/Interpreters/Aggregator.cpp [hunk 10]
```diff
@@ -27,12 +30,236 @@
...
+
+    using Cache = DB::LRUCache<UInt64, Entry>;
+    using CachePtr = std::shared_ptr<Cache>;
+    using Params = DB::Aggregator::Params::StatsCollectingParams;
+
+    /// Collection and use of the statistics should be enabled.
+    std::optional<Entry> getSizeHint(const Params & params)
+    {
+        if (!params.isCollectionAndUseEnabled())
+            throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Collection and use of the statistics should be enabled.");
```
Comment (vdimir) [inline src/Interpreters/Aggregator.cpp:65]: Maybe just return null and don't check `isCollectionAndUseEnabled` on caller side?

File: src/Interpreters/Aggregator.cpp [hunk 11]
```diff
@@ -27,12 +30,236 @@
...
+void updateStatistics(const DB::ManyAggregatedDataVariants & data_variants, const DB::Aggregator::Params::StatsCollectingParams & params)
+{
+    if (!params.isCollectionAndUseEnabled())
+        throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Collection and use of the statistics should be enabled.");
+
+    std::vector<size_t> sizes(data_variants.size());
+    for (size_t i = 0; i < data_variants.size(); ++i)
+        sizes[i] = data_variants[i]->size();
+    const auto median_size = sizes.begin() + sizes.size() / 2; // not precisely though...
+    std::nth_element(sizes.begin(), median_size, sizes.end());
```
Comment (vdimir) [inline src/Interpreters/Aggregator.cpp:215]: Why we do use median (but not let's say max)? Actually I'm not familiar with `AggregatedDataVariants` and didn't get why we have *many* `data_variants`, could you explain it to me?

**Review by vdimir — APPROVED** [review]
LGTM, but maybe @kitaisreal want take another look

**Review by kitaisreal — APPROVED** [review]
Looks good, need minor clarifications, check performance tests and ready to be merge.

**Comment by kitaisreal** [comment]
Performance tests looks good. We can additionally investigate why prellocating does not work good for StringHashTable in some separate pull request.

File: src/Interpreters/Aggregator.cpp (hunk 10 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:65]: I prefer straightforward code )

File: src/Interpreters/Aggregator.cpp (hunk 11 above)
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:215]: there is usually a separate variant for each thread of aggregation. [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/AggregatingStep.cpp#L155) is a relevant place in the code.
median seems to be the correct choice from statisctical considerations and experiments also agree with that.

File: src/Common/ProfileEvents.cpp [hunk 12]
```diff
@@ -284,6 +284,9 @@
     \
     M(MainConfigLoads, "Number of times the main configuration was reloaded.") \
     \
+    M(HashTablesPreallocatedElements, "How many elements were preallocated in hash tables for aggregation.") \
```
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Common/ProfileEvents.cpp]: it is mostly for testing and could be useful in debuging since all the added traces are of the 'debug' level

File: src/Interpreters/Aggregator.cpp [hunk 13]
```diff
@@ -27,12 +30,236 @@
...
+    }
+
+    /// Collection and use of the statistics should be enabled.
+    void update(size_t sum_of_sizes, size_t median_size, const Params & params)
+    {
+        if (!params.isCollectionAndUseEnabled())
+            throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Collection and use of the statistics should be enabled.");
+
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
```
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:89]: that's true, but it is nice to have operations with cache sequential

**Comment by nickitat** [comment]
> Performance tests looks good. We can additionally investigate why prellocating does not work good for StringHashTable in some separate pull request.

my understanding is that it is because StringHT is 4 HT under the hood for different size classes and unless the the distribution of input strings between sizes classes is uniform preallocation will not make much difference.

**Comment by alexey-milovidov** [comment]
Performance numbers are really great :+1:

File: tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh [hunk 14]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
```
Comment (kitaisreal) [inline tests/queries/0_stateless/02151_hash_table_sizes_stats_distributed.sh:1]: @nickitat need to fix this test.

**Comment by nickitat** [comment]
failures look irrelevant, but lets rerun just in case

File: src/Interpreters/Aggregator.cpp [hunk 15]
```diff
@@ -27,12 +30,236 @@
...
+        if (auto hint = getHashTablesStatistics().getSizeHint(stats_collecting_params))
+        {
+            const auto max_threads = params.group_by_two_level_threshold != 0 ? std::max(params.max_threads, 1ul) : 1;
+            const auto lower_limit = hint->sum_of_sizes / max_threads;
+            const auto upper_limit = stats_collecting_params.max_size_to_preallocate_for_aggregation / max_threads;
+            const auto adjusted = std::min(std::max(lower_limit, hint->median_size), upper_limit);
+            if (worthConvertToTwoLevel(
+                    params.group_by_two_level_threshold,
+                    hint->sum_of_sizes,
+                    /*group_by_two_level_threshold_bytes*/ 0,
```
Comment (Alex-Cheng) [inline src/Interpreters/Aggregator.cpp:194]: Does it mean the two-level aggregate never be used? why?

Comment (nickitat) [reply src/Interpreters/Aggregator.cpp:194]: it will be activated if `sum_of_sizes >= params.group_by_two_level_threshold`

**Comment by Nathaniel-Han** [comment]
Hi guys @nickitat , have you thought about apply the similar strategy for the hash join? I am wondering the improvement it can bring to the hash join.

**Comment by alexey-milovidov** [comment]
@Nathaniel-Han it should help and worth trying,
but it will help to a less extent, because more time is spent by moving data between blocks than by resizing the hash table.
